### ISSUE TITLE
aws_iam_policy_document renders Principals incorrectly

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#2216]
_This issue was originally opened by @kerbou as hashicorp/terraform#16590. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

Hi all

I'm using the aws_iam_policy_document data source to create an IAM policy for my managed AWS ElasticSearch cluster for managing anonymous access. However the returned output rendered by "data.aws_iam_policy_document.es_cluster.json" does not produce the desired output blocking otherwise valid requests

**Version: Terraform v0.10.7**

A bare-bone example of my policy data source (
 ```
data "aws_iam_policy_document" "es_cluster" {
  statement {
    actions = [
      "es:*",
    ]
    principals {
      type        = "AWS"
      identifiers = ["*"]
    }		
    resources = [
      " (some ARN) "
    ]
	condition {
      test     = "IpAddress"
      variable = "aws:SourceIp"
      values = [
		"1.1.1.1"
      ]
    }
  }
}
```

### Actual Behavior
The data source produces the following policy using data.aws_iam_policy_document.elasticsearch_cluster.json:

Notice the rendering of the principal 

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": "*",
      "Action": "es:*",
      "Resource": " (some ARN)",
      "Condition": {
        "IpAddress": {
          "aws:SourceIp": "1.1.1.1"
        }
      }
    }
  ]
}
```
Result: Requests to my ES cluster is blocked for anonymous users

### Expected Behavior
Changing the "principal" array by explicitly declaring "AWS" as a principal allows my anonymous requests to pass through as expected:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "es:*",
      "Resource": " (some ARN)",
      "Condition": {
        "IpAddress": {
          "aws:SourceIp": "1.1.1.1"
        }
      }
    }
  ]
}
```
In my view the rendering of the Principal is what's causing problems. I've tried a few tricks but haven't been able so far cheating the rendering engine into producing the desired output.

Is there a known workaround that I can use?

Best regards

P.S. - Kudos to the Terraform team and community by the way, the product is indeed great. This is the first time it's been causing me problems that wasn't entirely PEBCAK or ReadTheFuckingManual related (not ruling out that this one could fall into one of those categories as well)    :o)

**Comment by jurajseffer** [comment]
@kerbou workaround would be to not use `statement` at all and just explicitly templating it, e.g. inside of `aws_elasticsearch_domain` you'd have 
```
access_policies = <<CONFIG
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "es:*",
      "Resource": " (some ARN)",
      "Condition": {
        "IpAddress": {
          "aws:SourceIp": "1.1.1.1"
        }
      }
    }
  ]
}
CONFIG
```

**Comment by kerbou** [comment]
Nope - no idea if the blank SID changes anything



Den lør. 31. mar. 2018 08.22 skrev Brandon Ramsey <notifications@github.com

**Comment by mangobatao** [comment]
@brandonramsey hey, did you get it work for the anonymous access by
```"Principal":"*"```

**Comment by bflad** [comment]
FYI this pull request is probably relevant here where we may no longer normalize `type = "AWS"` and `identifier = "*"` to `"Principal": "*"` in the `aws_iam_policy_document` data source: https://github.com/terraform-providers/terraform-provider-aws/pull/4248

**Comment by bflad** [comment]
#4248 has been merged and will release with v1.19.0 of the AWS provider, which will now correctly handle the distinction between `type = "AWS"` and `type = "*"` to properly allow rendering `"Principal": {"AWS": "*"}` versus `"Principal": "*"` (respectively).

**Comment by bflad** [comment]
This has been released in [version 1.19.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1190-may-16-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
